File: robotics_planning/planners.py
"""Classical graph-search planners for occupancy grids."""

from __future__ import annotations

from dataclasses import dataclass
import heapq
from collections import deque
from math import hypot
from typing import Callable

from .grid import Cell, GridMap

Heuristic = Callable[[Cell, Cell], float]


@dataclass(frozen=True)
class SearchResult:
    """Planner output and metrics for one search run."""

    algorithm: str
    path: tuple[Cell, ...] | None
    expanded: tuple[Cell, ...]
    cost: float | None

    @property
    def found(self) -> bool:
        return self.path is not None

    @property
    def path_length(self) -> int | None:
        return None if self.path is None else max(0, len(self.path) - 1)


def manhattan(a: Cell, b: Cell) -> float:
    return float(abs(a[0] - b[0]) + abs(a[1] - b[1]))


def euclidean(a: Cell, b: Cell) -> float:
    return hypot(a[0] - b[0], a[1] - b[1])


def reconstruct_path(
    parent: dict[Cell, Cell], start: Cell, goal: Cell
) -> tuple[Cell, ...] | None:
    """Follow predecessor links from goal back to start, then reverse.

    Search algorithms store one predecessor per discovered cell rather than
    copying a full path into every queue entry. Reversing the recovered chain
    produces the start-to-goal path expected by the benchmark.
    """

    if start == goal:
        return (start,)
    if goal not in parent:
        return None

    reverse_path = [goal]
    current = goal
    while current != start:
        current = parent[current]
        reverse_path.append(current)
    reverse_path.reverse()
    return tuple(reverse_path)


def _validate_endpoints(grid: GridMap, start: Cell, goal: Cell) -> None:
    if not grid.is_free(start):
        raise ValueError(f"Start is not a free cell: {start}")
    if not grid.is_free(goal):
        raise ValueError(f"Goal is not a free cell: {goal}")


def _result(
    algorithm: str,
    parent: dict[Cell, Cell],
    start: Cell,
    goal: Cell,
    expanded: list[Cell],
    grid: GridMap,
) -> SearchResult:
    """Convert common search bookkeeping into the public result structure.

    The expanded sequence is kept for algorithm comparisons. Path cost is
    recomputed from the returned path so weighted cell-entry costs are
    included consistently for every planner.
    """
    path = reconstruct_path(parent, start, goal)
    if path is None:
        return SearchResult(algorithm, None, tuple(expanded), None)
    cost = sum(grid.cost_to_enter(cell) for cell in path[1:])
    return SearchResult(algorithm, path, tuple(expanded), cost)


class BFSPlanner:
    name = "BFS"

    def plan(self, grid: GridMap, start: Cell, goal: Cell) -> SearchResult:
        """Search in layers using a FIFO queue.

        With unit edge costs, the first visit to the goal has minimum step
        count. This baseline does not account for weighted cell costs.
        """
        _validate_endpoints(grid, start, goal)
        # ``frontier`` is the wavefront of cells whose neighbors have not yet
        # been inspected. ``discovered`` is marked at enqueue time, so the
        # same cell is never inserted twice through two equal-length routes.
        frontier = deque([start])
        discovered = {start}
        parent: dict[Cell, Cell] = {}
        expanded: list[Cell] = []

        while frontier:
            current = frontier.popleft()
            expanded.append(current)
            if current == goal:
                break
            # FIFO order visits all cells at distance d before distance d+1;
            # ``parent`` records the first shortest-step predecessor.
            for neighbor in grid.neighbors(current):
                if neighbor not in discovered:
                    discovered.add(neighbor)
                    parent[neighbor] = current
                    frontier.append(neighbor)

        return _result(self.name, parent, start, goal, expanded, grid)


class DijkstraPlanner:
    name = "Dijkstra"

    def plan(self, grid: GridMap, start: Cell, goal: Cell) -> SearchResult:
        """Expand the currently cheapest known cost-to-reach cell.

        Heap entries can become stale after a shorter route is found. The
        distance check discards those entries without requiring decrease-key.
        """
        _validate_endpoints(grid, start, goal)
        counter = 0
        # Heap entries are (known cost, insertion order, cell). The counter
        # makes equal-cost ordering deterministic without comparing cells.
        frontier: list[tuple[float, int, Cell]] = [(0.0, counter, start)]
        distances = {start: 0.0}
        parent: dict[Cell, Cell] = {}
        expanded: list[Cell] = []

        while frontier:
            distance, _, current = heapq.heappop(frontier)
            # A better route may have been found after this entry was pushed;
            # ignore the obsolete copy instead of expanding it again.
            if distance != distances.get(current):
                continue
            expanded.append(current)
            if current == goal:
                break
            for neighbor in grid.neighbors(current):
                # Relax the edge: replace the predecessor only when entering
                # this neighbor through the current cell is cheaper.
                new_distance = distance + grid.cost_to_enter(neighbor)
                if new_distance < distances.get(neighbor, float("inf")):
                    distances[neighbor] = new_distance
                    parent[neighbor] = current
                    counter += 1
                    heapq.heappush(frontier, (new_distance, counter, neighbor))

        return _result(self.name, parent, start, goal, expanded, grid)


class GreedyBestFirstPlanner:
    name = "Greedy"

    def __init__(self, heuristic: Heuristic = manhattan) -> None:
        self.heuristic = heuristic

    def plan(self, grid: GridMap, start: Cell, goal: Cell) -> SearchResult:
        """Prioritize cells that appear closest to the goal.

        Greedy search is a speed-oriented comparison, not an optimal planner:
        it ignores the cost already spent reaching the current cell.
        """
        _validate_endpoints(grid, start, goal)
        counter = 0
        # Greedy priority is h(n) only. ``discovered`` prevents cycles, but
        # unlike Dijkstra/A*, this search does not reopen a cell on a cheaper
        # route because path optimality is not its objective.
        frontier: list[tuple[float, int, Cell]] = [
            (self.heuristic(start, goal), counter, start)
        ]
        discovered = {start}
        parent: dict[Cell, Cell] = {}
        expanded: list[Cell] = []

        while frontier:
            _, _, current = heapq.heappop(frontier)
            expanded.append(current)
            if current == goal:
                break
            for neighbor in grid.neighbors(current):
                if neighbor not in discovered:
                    discovered.add(neighbor)
                    parent[neighbor] = current
                    counter += 1
                    priority = self.heuristic(neighbor, goal)
                    # Only the estimated remaining distance determines the
                    # queue order; the traveled cost is intentionally ignored.
                    heapq.heappush(frontier, (priority, counter, neighbor))

        return _result(self.name, parent, start, goal, expanded, grid)


class AStarPlanner:
    name = "A*"

    def __init__(
        self,
        heuristic: Heuristic = manhattan,
        prefer_goal_on_ties: bool = False,
    ) -> None:
        self.heuristic = heuristic
        self.prefer_goal_on_ties = prefer_goal_on_ties

    def plan(self, grid: GridMap, start: Cell, goal: Cell) -> SearchResult:
        """Minimize ``cost_so_far + heuristic`` with optional tie-breaking.

        The heuristic changes queue order, while ``distances`` controls route
        replacement. With an admissible heuristic, A* preserves shortest-path
        cost; the optional goal tie-break only selects among equal priorities.
        """
        _validate_endpoints(grid, start, goal)
        counter = 0
        start_heuristic = self.heuristic(start, goal)
        start_tie_break = start_heuristic if self.prefer_goal_on_ties else 0.0
        # A* entries are (f, tie_break, insertion_order, cell, g), where
        # f=g+h is the estimated total path cost and g is retained to reject
        # stale heap entries. The extra fields only make ties reproducible.
        frontier: list[tuple[float, float, int, Cell, float]] = [
            (start_heuristic, start_tie_break, counter, start, 0.0)
        ]
        distances = {start: 0.0}
        parent: dict[Cell, Cell] = {}
        expanded: list[Cell] = []

        while frontier:
            _, _, _, current, queued_cost = heapq.heappop(frontier)
            # ``current`` may have an old, larger g value in the heap after a
            # shorter path updated ``distances``. It must not be expanded.
            if queued_cost != distances.get(current):
                continue
            current_cost = distances[current]
            expanded.append(current)
            if current == goal:
                break
            for neighbor in grid.neighbors(current):
                # Relax this edge using the true accumulated cost. The
                # heuristic is used only to prioritize the resulting entry.
                new_cost = current_cost + grid.cost_to_enter(neighbor)
                if new_cost < distances.get(neighbor, float("inf")):
                    distances[neighbor] = new_cost
                    parent[neighbor] = current
                    counter += 1
                    heuristic_value = self.heuristic(neighbor, goal)
                    priority = new_cost + heuristic_value
                    # With an admissible h, f preserves A* optimality. The
                    # optional tie-break prefers smaller h among equal f and
                    # therefore tends to expand cells closer to the goal.
                    tie_break = (
                        heuristic_value if self.prefer_goal_on_ties else 0.0
                    )
                    heapq.heappush(
                        frontier,
                        (priority, tie_break, counter, neighbor, new_cost),
                    )

        return _result(self.name, parent, start, goal, expanded, grid)


class DFSBacktrackingPlanner:
    """Depth-first maze solver for comparison and educational purposes.

    The implementation uses an explicit stack rather than Python recursion so
    that scaling experiments can include larger grids without hitting the
    interpreter recursion limit.
    """

    name = "DFS backtracking"

    def plan(self, grid: GridMap, start: Cell, goal: Cell) -> SearchResult:
        """Explore one branch at a time with an explicit backtracking stack.

        Each stack entry stores the next neighbor index to inspect. This is
        recursive DFS expressed iteratively, avoiding Python recursion limits
        in scaling experiments.
        """
        _validate_endpoints(grid, start, goal)
        visited: set[Cell] = set()
        expanded: list[Cell] = []
        path: list[Cell] = []

        visited.add(start)
        expanded.append(start)
        path.append(start)
        # Each stack tuple is (cell, fixed neighbor list, next index). Keeping
        # the index is the iterative equivalent of returning from a recursive
        # DFS call and continuing with the next sibling.
        stack: list[tuple[Cell, tuple[Cell, ...], int]] = [
            (start, grid.neighbors(start), 0)
        ]

        while stack:
            current, neighbors, next_index = stack[-1]
            if current == goal:
                break
            if next_index >= len(neighbors):
                # Every branch below this cell was exhausted: backtrack one
                # level and remove it from the current candidate path.
                stack.pop()
                path.pop()
                continue

            stack[-1] = (current, neighbors, next_index + 1)
            neighbor = neighbors[next_index]
            if neighbor in visited:
                continue

            # DFS marks on discovery so cycles cannot re-enter an old branch.
            visited.add(neighbor)
            expanded.append(neighbor)
            path.append(neighbor)
            stack.append((neighbor, grid.neighbors(neighbor), 0))

        if not stack or not path or path[-1] != goal:
            return SearchResult(self.name, None, tuple(expanded), None)
        cost = sum(grid.cost_to_enter(cell) for cell in path[1:])
        return SearchResult(self.name, tuple(path), tuple(expanded), cost)

File: cpp/src/planners.cpp
#include "robotics_planning/planners.hpp"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <functional>
#include <limits>
#include <queue>
#include <stdexcept>

namespace robotics_planning {
namespace {

struct QueueEntry {
    // ``priority`` is h for Greedy and g+h for A*. ``g_cost`` is stored
    // separately so stale entries can be rejected after a better route is
    // discovered for the same cell.
    double priority;
    std::size_t sequence;
    Cell cell;
    double g_cost;
};

struct MinQueue {
    bool operator()(const QueueEntry& left, const QueueEntry& right) const {
        if (left.priority != right.priority) {
            return left.priority > right.priority;
        }
        return left.sequence > right.sequence;
    }
};

void validate_endpoints(const GridMap& grid, Cell start, Cell goal) {
    if (!grid.is_free(start)) {
        throw std::invalid_argument("Start is not a free cell");
    }
    if (!grid.is_free(goal)) {
        throw std::invalid_argument("Goal is not a free cell");
    }
}

double path_cost(const GridMap& grid, const std::vector<Cell>& path) {
    // The cost of a path is the sum of costs for entered cells; the start
    // cell is not charged because the robot is already there.
    double total = 0.0;
    for (std::size_t index = 1; index < path.size(); ++index) {
        total += grid.cost_to_enter(path[index]);
    }
    return total;
}

SearchResult make_result(
    const std::string& algorithm,
    const GridMap& grid,
    const std::vector<Cell>& parent,
    Cell start,
    Cell goal,
    std::vector<Cell> expanded
) {
    // All planners share path reconstruction and cost reporting so that
    // differences in the benchmark come from search order, not accounting.
    SearchResult result;
    result.algorithm = algorithm;
    result.expanded = std::move(expanded);
    result.path = reconstruct_path(parent, start, goal);
    result.found = !result.path.empty();
    if (result.found) {
        result.cost = path_cost(grid, result.path);
    }
    return result;
}

}  // namespace

double heuristic(
    const GridMap& grid,
    Cell current,
    Cell goal,
    HeuristicKind kind
) {
    const double dx = static_cast<double>(grid.x_of(current) - grid.x_of(goal));
    const double dy = static_cast<double>(grid.y_of(current) - grid.y_of(goal));
    if (kind == HeuristicKind::Euclidean) {
        return std::sqrt(dx * dx + dy * dy);
    }
    return std::abs(dx) + std::abs(dy);
}

std::vector<Cell> reconstruct_path(
    const std::vector<Cell>& parent,
    Cell start,
    Cell goal
) {
    // Search stores one predecessor per cell. Walk from the goal backward,
    // then reverse the vector to recover the command order start -> goal.
    if (start == goal) {
        return {start};
    }
    if (goal < 0 || goal >= static_cast<Cell>(parent.size()) || parent[goal] < 0) {
        return {};
    }

    std::vector<Cell> reverse_path;
    reverse_path.push_back(goal);
    Cell current = goal;
    while (current != start) {
        if (current < 0 || current >= static_cast<Cell>(parent.size())) {
            return {};
        }
        current = parent[current];
        if (current < 0) {
            return {};
        }
        reverse_path.push_back(current);
    }

    std::reverse(reverse_path.begin(), reverse_path.end());
    return reverse_path;
}

SearchResult BFSPlanner::plan(const GridMap& grid, Cell start, Cell goal) const {
    validate_endpoints(grid, start, goal);
    // FIFO order creates a breadth-wise wavefront. Because every move has
    // unit cost, the first discovered route to a cell has minimum steps.
    std::queue<Cell> frontier;
    std::vector<std::uint8_t> discovered(static_cast<std::size_t>(grid.cell_count()), 0U);
    std::vector<Cell> parent(static_cast<std::size_t>(grid.cell_count()), -1);
    std::vector<Cell> expanded;

    frontier.push(start);
    discovered[static_cast<std::size_t>(start)] = 1U;

    while (!frontier.empty()) {
        const Cell current = frontier.front();
        frontier.pop();
        expanded.push_back(current);
        if (current == goal) {
            break;
        }

        for (const Cell neighbor : grid.neighbors4(current)) {
            // Mark at enqueue time, not dequeue time, to avoid duplicate
            // queue entries and to preserve the first shortest predecessor.
            if (neighbor >= 0 && discovered[static_cast<std::size_t>(neighbor)] == 0U) {
                discovered[static_cast<std::size_t>(neighbor)] = 1U;
                parent[static_cast<std::size_t>(neighbor)] = current;
                frontier.push(neighbor);
            }
        }
    }

    return make_result("BFS", grid, parent, start, goal, std::move(expanded));
}

SearchResult DijkstraPlanner::plan(const GridMap& grid, Cell start, Cell goal) const {
    validate_endpoints(grid, start, goal);
    const double infinity = std::numeric_limits<double>::infinity();
    std::vector<double> distance(static_cast<std::size_t>(grid.cell_count()), infinity);
    std::vector<Cell> parent(static_cast<std::size_t>(grid.cell_count()), -1);
    std::vector<Cell> expanded;
    // The smallest known path cost is expanded first. The priority queue may
    // contain multiple entries for one cell because C++ priority_queue has no
    // decrease-key operation.
    std::priority_queue<QueueEntry, std::vector<QueueEntry>, MinQueue> frontier;
    std::size_t sequence = 0;

    distance[static_cast<std::size_t>(start)] = 0.0;
    frontier.push({0.0, sequence++, start, 0.0});

    while (!frontier.empty()) {
        const QueueEntry entry = frontier.top();
        frontier.pop();
        // A later relaxation may have lowered this cell's distance. The old
        // entry is stale and must not produce more expansions.
        if (entry.g_cost > distance[static_cast<std::size_t>(entry.cell)]) {
            continue;
        }

        expanded.push_back(entry.cell);
        if (entry.cell == goal) {
            break;
        }

        for (const Cell neighbor : grid.neighbors4(entry.cell)) {
            if (neighbor < 0) {
                continue;
            }
            const double new_cost = entry.g_cost + grid.cost_to_enter(neighbor);
            auto& known_cost = distance[static_cast<std::size_t>(neighbor)];
            // Relax the edge only when this route improves the best known g.
            if (new_cost < known_cost) {
                known_cost = new_cost;
                parent[static_cast<std::size_t>(neighbor)] = entry.cell;
                frontier.push({new_cost, sequence++, neighbor, new_cost});
            }
        }
    }

    return make_result("Dijkstra", grid, parent, start, goal, std::move(expanded));
}

SearchResult GreedyBestFirstPlanner::plan(
    const GridMap& grid,
    Cell start,
    Cell goal
) const {
    validate_endpoints(grid, start, goal);
    std::vector<std::uint8_t> discovered(static_cast<std::size_t>(grid.cell_count()), 0U);
    std::vector<Cell> parent(static_cast<std::size_t>(grid.cell_count()), -1);
    std::vector<Cell> expanded;
    // Greedy priority is h(n), the estimated remaining distance only. It can
    // reach the goal quickly but may ignore an expensive or long detour.
    std::priority_queue<QueueEntry, std::vector<QueueEntry>, MinQueue> frontier;
    std::size_t sequence = 0;

    discovered[static_cast<std::size_t>(start)] = 1U;
    frontier.push({heuristic(grid, start, goal, heuristic_kind_), sequence++, start, 0.0});

    while (!frontier.empty()) {
        const QueueEntry entry = frontier.top();
        frontier.pop();
        expanded.push_back(entry.cell);
        if (entry.cell == goal) {
            break;
        }

        for (const Cell neighbor : grid.neighbors4(entry.cell)) {
            if (neighbor >= 0 && discovered[static_cast<std::size_t>(neighbor)] == 0U) {
                discovered[static_cast<std::size_t>(neighbor)] = 1U;
                parent[static_cast<std::size_t>(neighbor)] = entry.cell;
                // Greedy marks cells once and does not reopen them through a
                // cheaper route; optimality is intentionally not guaranteed.
                frontier.push({
                    heuristic(grid, neighbor, goal, heuristic_kind_),
                    sequence++,
                    neighbor,
                    0.0,
                });
            }
        }
    }

    return make_result("Greedy", grid, parent, start, goal, std::move(expanded));
}

SearchResult AStarPlanner::plan(const GridMap& grid, Cell start, Cell goal) const {
    validate_endpoints(grid, start, goal);
    const double infinity = std::numeric_limits<double>::infinity();
    std::vector<double> distance(static_cast<std::size_t>(grid.cell_count()), infinity);
    std::vector<Cell> parent(static_cast<std::size_t>(grid.cell_count()), -1);
    std::vector<Cell> expanded;
    // A* ranks entries by f(n) = g(n) + h(n). The heuristic only changes the
    // order in which candidates are explored; distance[] remains the source
    // of truth for route replacement.
    std::priority_queue<QueueEntry, std::vector<QueueEntry>, MinQueue> frontier;
    std::size_t sequence = 0;

    distance[static_cast<std::size_t>(start)] = 0.0;
    frontier.push({
        heuristic(grid, start, goal, heuristic_kind_),
        sequence++,
        start,
        0.0,
    });

    while (!frontier.empty()) {
        const QueueEntry entry = frontier.top();
        frontier.pop();
        // Ignore an obsolete copy whose stored g is larger than the current
        // best distance for this cell.
        if (entry.g_cost > distance[static_cast<std::size_t>(entry.cell)]) {
            continue;
        }

        expanded.push_back(entry.cell);
        if (entry.cell == goal) {
            break;
        }

        for (const Cell neighbor : grid.neighbors4(entry.cell)) {
            if (neighbor < 0) {
                continue;
            }
            const double new_cost = entry.g_cost + grid.cost_to_enter(neighbor);
            auto& known_cost = distance[static_cast<std::size_t>(neighbor)];
            // This is the relaxation step. With an admissible heuristic,
            // stopping when the goal is popped preserves optimal path cost.
            if (new_cost < known_cost) {
                known_cost = new_cost;
                parent[static_cast<std::size_t>(neighbor)] = entry.cell;
                frontier.push({
                    new_cost + heuristic(grid, neighbor, goal, heuristic_kind_),
                    sequence++,
                    neighbor,
                    new_cost,
                });
            }
        }
    }

    return make_result("A*", grid, parent, start, goal, std::move(expanded));
}

SearchResult DFSBacktrackingPlanner::plan(
    const GridMap& grid,
    Cell start,
    Cell goal
) const {
    validate_endpoints(grid, start, goal);
    std::vector<std::uint8_t> visited(static_cast<std::size_t>(grid.cell_count()), 0U);
    std::vector<Cell> parent(static_cast<std::size_t>(grid.cell_count()), -1);
    std::vector<Cell> expanded;

    // The lambda expresses recursive DFS. ``parent`` is assigned immediately
    // before descending, so reconstruct_path can recover the active branch if
    // the goal is found.
    std::function<bool(Cell)> visit = [&](Cell current) {
        visited[static_cast<std::size_t>(current)] = 1U;
        expanded.push_back(current);
        if (current == goal) {
            return true;
        }

        for (const Cell neighbor : grid.neighbors4(current)) {
            if (neighbor >= 0 && visited[static_cast<std::size_t>(neighbor)] == 0U) {
                // Marking before recursion prevents cycles in undirected grid
                // edges and makes each cell expand at most once.
                parent[static_cast<std::size_t>(neighbor)] = current;
                if (visit(neighbor)) {
                    return true;
                }
            }
        }
        return false;
    };

    visit(start);
    return make_result(
        "DFS backtracking",
        grid,
        parent,
        start,
        goal,
        std::move(expanded)
    );
}

}  // namespace robotics_planning

